File: quicksort/partition.js

function partition(a, lo, hi) {
  let i = lo;
  let j = hi + 1;
  let v = a[i]

  while(true) {
    while(a[++i] < v) {
      if(i === hi) { break; }
    }
    while(v < a[--j]) {
      if(j === lo) { break; }
    }
    if(i >= j) { break; }
    swap(a, i, j)
  }
  swap(a, lo, j)
  return j
}

function swap(arr, a, b) {
  let temp = arr[a];
  arr[a] = arr[b];
  arr[b] = temp;
}

let samp_a = ['k', 'r', 'a', 't', 'e', 'l', 'e', 'p', 'u', 'i', 'm', 'q', 'c', 'x', 'o', 's']
let lo = 0
let hi = 15

console.log(partition(samp_a, lo, hi))

File: quicksort/partition.rb
# frozen_string_literal: true

require 'pry'

def partition(a, lo, hi)
  i = lo
  j = hi + 1
  v = a[i]

  while true
    while a[i += 1] < v
      break if i == hi
    end
    while v < a[j -= 1]
      break if j == lo
    end
    break if i >= j

    # Parallel Assignment
    a[i], a[j] = a[j], a[i]
  end

  a[lo], a[j] = a[j], a[lo]
  j
end

samp_a = ['k', 'r', 'a', 't', 'e', 'l', 'e', 'p', 'u', 'i', 'm', 'q', 'c', 'x', 'o', 's']
lo = 0
hi = 15

partition(samp_a, lo, hi)
